File: interface/src/lib/services/api/integrations/ImuApiService.ts
import { createApiClient, type ApiClientOptions } from '$lib/utils';

export interface ImuVector3 {
	x: number;
	y: number;
	z: number;
}

export interface ImuSettings {
	ui_monitor_enabled: boolean;
	alarm_monitor_enabled: boolean;
	orientation_baseline_valid: boolean;
	orientation_baseline: ImuVector3;
	baseline_calibrated_at: number;
	calibration_revision: number;
	tilt_threshold_deg: number;
	tilt_hysteresis_deg: number;
	tilt_hold_ms: number;
	tilt_clear_hold_ms: number;
	accel_delta_threshold_g: number;
}

export interface ImuConsumerState {
	desired: boolean;
	running: boolean;
}

interface ImuSample {
	ax: number;
	ay: number;
	az: number;
	gx: number;
	gy: number;
	gz: number;
}

export interface ImuMetrics {
	accel_magnitude_g: number;
	accel_delta_g: number;
	gyro_magnitude_dps: number;
	baseline_valid: boolean;
	orientation_baseline: ImuVector3;
	tilt_deg: number | null;
	sample: ImuSample;
}

export interface ImuAlarmStatus {
	enabled: boolean;
	sample_fresh: boolean;
	baseline_valid: boolean;
	triggered: boolean;
	pending_trigger: boolean;
	pending_clear: boolean;
	reason: 'none' | 'tilt' | 'shock' | 'stale' | 'no_baseline' | 'unavailable' | 'unknown';
	tilt_deg: number | null;
	accel_delta_g: number;
	trigger_value: number;
	trigger_hold_elapsed_ms: number;
	clear_hold_elapsed_ms: number;
}

export interface ImuStatus {
	initialized: boolean;
	running: boolean;
	transition_in_progress: boolean;
	desired_mask: number;
	running_mask: number;
	desired_consumers: string;
	running_consumers: string;
	last_start_error: string;
	last_start_attempt_ms: number;
	last_start_duration_ms: number;
	next_retry_ms: number;
	retry_pending: boolean;
	sample_fresh: boolean;
	sample_age_ms: number | null;
	last_sample_ms: number | null;
	consumers: Record<
		'airmouse_movement' | 'airmouse_click' | 'auto_rotate' | 'alarm' | 'ui_monitor',
		ImuConsumerState
	>;
	alarm: ImuAlarmStatus;
	metrics: ImuMetrics;
}

export interface ImuCalibrationResult {
	ok: boolean;
	status: string;
	sample_count: number;
	accel_magnitude_mean: number;
	accel_magnitude_variance: number;
	orientation_baseline: ImuVector3;
}

export interface ImuResetResult {
	ok: boolean;
	status: string;
}

export const DEFAULT_IMU_SETTINGS: ImuSettings = {
	ui_monitor_enabled: false,
	alarm_monitor_enabled: false,
	orientation_baseline_valid: false,
	orientation_baseline: { x: 0, y: 0, z: 1 },
	baseline_calibrated_at: 0,
	calibration_revision: 0,
	tilt_threshold_deg: 30,
	tilt_hysteresis_deg: 5,
	tilt_hold_ms: 750,
	tilt_clear_hold_ms: 1500,
	accel_delta_threshold_g: 0.35
};

export class ImuApiService {
	private client;
	private static readonly GET_TIMEOUT_MS = 5000;
	private static readonly UPDATE_TIMEOUT_MS = 8000;
	private static readonly CALIBRATE_TIMEOUT_MS = 20000;

	constructor(options: ApiClientOptions) {
		this.client = createApiClient(options);
	}

	async getStatus(signal?: AbortSignal): Promise<ImuStatus> {
		return this.client.get<ImuStatus>('/api/imu/status', {
			signal: signal
				? AbortSignal.any([signal, AbortSignal.timeout(ImuApiService.GET_TIMEOUT_MS)])
				: AbortSignal.timeout(ImuApiService.GET_TIMEOUT_MS)
		});
	}

	async getSettings(): Promise<ImuSettings> {
		return this.client.get<ImuSettings>('/api/imu/settings', {
			signal: AbortSignal.timeout(ImuApiService.GET_TIMEOUT_MS)
		});
	}

	async updateSettings(settings: ImuSettings): Promise<ImuSettings> {
		return this.client.post<ImuSettings>('/api/imu/settings', settings, {
			signal: AbortSignal.timeout(ImuApiService.UPDATE_TIMEOUT_MS)
		});
	}

	async calibrateOrientation(): Promise<ImuCalibrationResult> {
		return this.client.post<ImuCalibrationResult>(
			'/api/imu/calibrate/orientation',
			{},
			{ signal: AbortSignal.timeout(ImuApiService.CALIBRATE_TIMEOUT_MS) }
		);
	}

	async resetOrientationBaseline(): Promise<ImuResetResult> {
		return this.client.post<ImuResetResult>(
			'/api/imu/reset/orientation-baseline',
			{},
			{ signal: AbortSignal.timeout(ImuApiService.UPDATE_TIMEOUT_MS) }
		);
	}
}

File: src/api/imu/ImuApiService.cpp
#include "ImuApiService.h"

#include "../../config/json/ConfigKeys.h"
#include "../../sensors/imu/ImuMath.h"
#include "../../sensors/imu/ImuRuntimeService.h"
#include "../../system/utils/json/JsonResponseWriter.h"

#include <utils/ResponseUtils.h>

#include <cmath>

namespace API {

namespace {
constexpr const char* kSettingsPath = "/api/imu/settings";
constexpr const char* kStatusPath = "/api/imu/status";
constexpr const char* kCalibrateOrientationPath = "/api/imu/calibrate/orientation";
constexpr const char* kUnavailableError = "imu/unavailable";

constexpr uint32_t consumerBit(IMU::Consumer consumer) {
    return 1u << static_cast<uint8_t>(consumer);
}

void writeConsumer(Utils::JsonResponseWriter& w,
                   const char* key,
                   IMU::Consumer consumer,
                   uint32_t desiredMask,
                   uint32_t runningMask) {
    const uint32_t bit = consumerBit(consumer);
    w.raw(","); w.string(key); w.raw(":{");
    w.string("desired"); w.raw(":"); w.value((desiredMask & bit) != 0);
    w.raw(","); w.key(CONFIG::Keys::kRunning); w.value((runningMask & bit) != 0);
    w.raw("}");
}

void writeVector(Utils::JsonResponseWriter& w, const IMU::ImuVector3& v) {
    w.raw("{");
    w.key(CONFIG::Keys::kX); w.value(v.x, 4);
    w.raw(","); w.key(CONFIG::Keys::kY); w.value(v.y, 4);
    w.raw(","); w.key(CONFIG::Keys::kZ); w.value(v.z, 4);
    w.raw("}");
}

}  // namespace

ImuApiService::ImuApiService(PsychicHttpServer* server,
                             SecurityManager* securityManager,
                             POWER::PowerManager* powerManager,
                             IMU::ImuRuntimeService* runtime)
    : BaseApiService(server, securityManager, powerManager, "api/imu"),
      _runtime(runtime) {
    if (_runtime) {
        _settingsEndpoint = std::make_unique<HttpEndpoint<RTC::ImuData>>(
            IMU::ImuRuntimeService::readState,
            IMU::ImuRuntimeService::updateState,
            _runtime,
            _server,
            kSettingsPath,
            _securityManager,
            AuthenticationPredicates::IS_ADMIN,
            AuthenticationPredicates::IS_AUTHENTICATED,
            nullptr,
            [this]() {
                if (_powerManager) {
                    _powerManager->notifyActivity(_activityTag);
                }
            });
    }
}

void ImuApiService::begin() {
    _server->on(kStatusPath, HTTP_GET, wrapAuth([this](PsychicRequest* request) {
        return handleStatus(request);
    }));

    if (_settingsEndpoint) {
        _settingsEndpoint->begin();
    }

    _server->on(kCalibrateOrientationPath, HTTP_POST, wrapAdmin([this](PsychicRequest* request) {
        return handleCalibrateOrientation(request);
    }));
}

esp_err_t ImuApiService::handleStatus(PsychicRequest* request) {
    if (!_runtime) {
        return Response::error(request, 503, kUnavailableError);
    }

    const IMU::ManagerStatus manager = _runtime->getManagerStatus();
    const IMU::ImuMetrics metrics = _runtime->getMetrics();

    char desiredConsumers[128];
    char runningConsumers[128];
    IMU::ImuManager::formatConsumers(manager.desiredMask, desiredConsumers, sizeof(desiredConsumers));
    IMU::ImuManager::formatConsumers(manager.runningMask, runningConsumers, sizeof(runningConsumers));

    Utils::JsonResponseWriter w(request->request());
    if (!w.beginResponse()) return ESP_FAIL;

    w.raw("{");
    w.key(CONFIG::Keys::kInitialized); w.value(manager.initialized);
    w.raw(","); w.key(CONFIG::Keys::kRunning); w.value(manager.initialized && manager.runningMask != 0);
    w.raw(","); w.key(CONFIG::Keys::kTransitionInProgress); w.value(manager.transitionInProgress);
    w.raw(","); w.key(CONFIG::Keys::kDesiredMask); w.value(static_cast<unsigned long>(manager.desiredMask));
    w.raw(","); w.key(CONFIG::Keys::kRunningMask); w.value(static_cast<unsigned long>(manager.runningMask));
    w.raw(","); w.key(CONFIG::Keys::kDesiredConsumers); w.string(desiredConsumers);
    w.raw(","); w.key(CONFIG::Keys::kRunningConsumers); w.string(runningConsumers);
    w.raw(","); w.key(CONFIG::Keys::kLastStartError);
    w.string(IMU::ImuManager::startErrorToString(manager.lastStartError));
    w.raw(","); w.key(CONFIG::Keys::kLastStartAttemptMs); w.value(static_cast<unsigned long>(manager.lastStartAttemptMs));
    w.raw(","); w.key(CONFIG::Keys::kLastStartDurationMs); w.value(static_cast<unsigned long>(manager.lastStartDurationMs));
    w.raw(","); w.key(CONFIG::Keys::kNextRetryMs); w.value(static_cast<unsigned long>(manager.nextRetryMs));
    w.raw(","); w.key(CONFIG::Keys::kRetryPending);
    w.value(manager.lastStartError == IMU::StartError::RetryPending ||
            manager.lastStartError == IMU::StartError::StartFailed);
    w.raw(","); w.key(CONFIG::Keys::kSampleFresh); w.value(metrics.sampleFresh);
    w.raw(","); w.key(CONFIG::Keys::kSampleAgeMs); w.value(static_cast<unsigned long>(metrics.sampleAgeMs));
    w.raw(","); w.key(CONFIG::Keys::kLastSampleMs); w.value(static_cast<unsigned long>(metrics.sample.timestampMs));

    w.raw(","); w.key(CONFIG::Keys::kConsumers);
    w.raw("{");
    w.string("airmouse_movement"); w.raw(":{");
    w.string("desired"); w.raw(":"); w.value((manager.desiredMask & consumerBit(IMU::Consumer::AirMouseMovement)) != 0);
    w.raw(","); w.key(CONFIG::Keys::kRunning); w.value((manager.runningMask & consumerBit(IMU::Consumer::AirMouseMovement)) != 0);
    w.raw("}");
    writeConsumer(w, "airmouse_click", IMU::Consumer::AirMouseClick, manager.desiredMask, manager.runningMask);
    writeConsumer(w, "auto_rotate", IMU::Consumer::AutoRotate, manager.desiredMask, manager.runningMask);
    writeConsumer(w, "alarm", IMU::Consumer::Alarm, manager.desiredMask, manager.runningMask);
    writeConsumer(w, "ui_monitor", IMU::Consumer::UiMonitor, manager.desiredMask, manager.runningMask);
    w.raw("}");

    w.raw(","); w.key(CONFIG::Keys::kMetrics);
    w.raw("{");
    w.key(CONFIG::Keys::kAccelMagnitudeG); w.value(metrics.accelMagnitudeG, 4);
    w.raw(","); w.key(CONFIG::Keys::kAccelDeltaG); w.value(metrics.accelDeltaG, 4);
    w.raw(","); w.key(CONFIG::Keys::kGyroMagnitudeDps); w.value(metrics.gyroMagnitudeDps, 2);
    w.raw(","); w.key(CONFIG::Keys::kBaselineValid); w.value(metrics.baselineValid);
    w.raw(","); w.key(CONFIG::Keys::kOrientationBaseline); writeVector(w, metrics.baseline);
    w.raw(","); w.key(CONFIG::Keys::kTiltDeg); w.value(metrics.tiltDeg, 2);
    w.raw(","); w.key(CONFIG::Keys::kSample);
    w.raw("{");
    w.key(CONFIG::Keys::kAx); w.value(metrics.sample.ax, 4);
    w.raw(","); w.key(CONFIG::Keys::kAy); w.value(metrics.sample.ay, 4);
    w.raw(","); w.key(CONFIG::Keys::kAz); w.value(metrics.sample.az, 4);
    w.raw(","); w.key(CONFIG::Keys::kGx); w.value(metrics.sample.gx, 4);
    w.raw(","); w.key(CONFIG::Keys::kGy); w.value(metrics.sample.gy, 4);
    w.raw(","); w.key(CONFIG::Keys::kGz); w.value(metrics.sample.gz, 4);
    w.raw("}");
    w.raw("}");
    w.raw("}");

    return w.finish() ? ESP_OK : ESP_FAIL;
}

esp_err_t ImuApiService::handleCalibrateOrientation(PsychicRequest* request) {
    if (!_runtime) {
        return Response::error(request, 503, kUnavailableError);
    }

    const IMU::OrientationCalibrationResult result = _runtime->calibrateOrientation();
    return Response::success(request, [&](JsonVariant& root) {
        root["ok"] = result.status == IMU::OrientationCalibrationStatus::Success;
        root[CONFIG::Keys::kStatus] = IMU::MATH::calibrationStatusToString(result.status);
        root["sample_count"] = result.sampleCount;
        root["accel_magnitude_mean"] = result.accelMagnitudeMean;
        root["accel_magnitude_variance"] = result.accelMagnitudeVariance;
        JsonObject baseline = root[CONFIG::Keys::kOrientationBaseline].to<JsonObject>();
        baseline[CONFIG::Keys::kX] = result.baseline.x;
        baseline[CONFIG::Keys::kY] = result.baseline.y;
        baseline[CONFIG::Keys::kZ] = result.baseline.z;
    });
}

}  // namespace API
